File: software engineering-fronted/frontend-vue-app/src/services/graph.ts
import request, { USE_MOCK } from '@/utils/request'
import { mockGraphApi } from './mock'

export interface GraphNode {
  id: number
  name: string
  description: string
  document_id: number
  category?: string
  // D3 force simulation properties
  x?: number
  y?: number
  vx?: number
  vy?: number
  fx?: number | null
  fy?: number | null
}

export interface GraphEdge {
  id: number
  source: number
  target: number
  relation_type: string
  description?: string
}

export interface GraphData {
  nodes: GraphNode[]
  edges: GraphEdge[]
  summary: {
    node_count: number
    edge_count: number
  }
}

export interface GraphBuildResult {
  build_id: number
  created_points: number
  created_relations: number
  chunk_count: number
  vector_count: number
  status: string
  message: string
}

// 获取图谱数据
export async function getGraphData(params?: { document_id?: number; keyword?: string; relation_type?: string }) {
  if (USE_MOCK) {
    return mockGraphApi.getGraphData(params) as Promise<any>
  }
  return request.get<GraphData>('/graph', { params })
}

// 从文档构建图谱
export async function buildGraph(document_ids: number[]) {
  if (USE_MOCK) {
    return mockGraphApi.buildGraph(document_ids) as Promise<any>
  }
  return request.post<GraphBuildResult>('/graph/build', { document_ids })
}

// 获取最近构建结果
export function getLatestBuild() {
  return request.get('/graph/build/latest')
}

// 构建历史记录
export function getBuildHistory(params?: { page?: number; size?: number }) {
  return request.get('/graph/build/history', { params })
}

// 更新知识点（节点）
export function updateKnowledgePoint(id: number, data: { name?: string; description?: string; category?: string }) {
  return request.put(`/knowledge/points/${id}`, data)
}

File: software engineering-back/internal/model/dto/response/graph.go
package response

// GraphDataResponse 知识图谱数据响应
type GraphDataResponse struct {
	Nodes   []GraphNode  `json:"nodes"`   // 节点列表
	Edges   []GraphEdge  `json:"edges"`   // 边列表
	Summary GraphSummary `json:"summary"` // 图谱统计摘要
}

// GraphNode 知识图谱节点
type GraphNode struct {
	ID          uint   `json:"id"`          // 节点ID
	Name        string `json:"name"`        // 节点名称
	Description string `json:"description"` // 节点描述
	DocumentID  uint   `json:"document_id"` // 关联的文档ID
	Category    string `json:"category"`    // 节点分类
}

// GraphEdge 知识图谱边（关系）
type GraphEdge struct {
	ID           uint   `json:"id"`             // 边ID
	Source       uint   `json:"source"`         // 源节点ID
	Target       uint   `json:"target"`         // 目标节点ID
	RelationType string `json:"relation_type"`  // 关系类型
	Description  string `json:"description"`    // 关系描述
}

// GraphSummary 知识图谱统计摘要
type GraphSummary struct {
	NodeCount int `json:"node_count"` // 节点数量
	EdgeCount int `json:"edge_count"` // 边数量
}

// BuildGraphResponse 构建知识图谱响应
type BuildGraphResponse struct {
	BuildID          uint   `json:"build_id"`          // 构建记录ID
	CreatedPoints    int    `json:"created_points"`    // 创建的知识点数量
	CreatedRelations int    `json:"created_relations"` // 创建的关系数量
	ChunkCount       int    `json:"chunk_count"`       // 文档分块数量
	VectorCount      int    `json:"vector_count"`      // 向量数量
	Status           string `json:"status"`            // 构建状态
	Message          string `json:"message"`           // 构建结果描述
}

// BuildHistoryResponse 构建历史响应
type BuildHistoryResponse struct {
	List      []BuildGraphResponse `json:"list"`       // 构建记录列表
	Total     int64                `json:"total"`      // 总记录数
	Page      int                  `json:"page"`       // 当前页码
	Size      int                  `json:"size"`       // 每页数量
	TotalPage int                  `json:"total_page"` // 总页数
}
